scipy/optimize/minpack.py

```diff
@@ -449,12 +449,24 @@ def leastsq(func, x0, args=(), Dfun=None, full_output=0,
         return (retval[0], info)
 
 
-def _general_function(params, xdata, ydata, function):
-    return function(xdata, *params) - ydata
+def _wrap_func(func, xdata, ydata, weights):
+    if weights is None:
+        def func_wrapped(params):
+            return func(xdata, *params) - ydata
+    else:
+        def func_wrapped(params):
+            return weights * (func(xdata, *params) - ydata)
+    return func_wrapped
 
 
-def _weighted_general_function(params, xdata, ydata, function, weights):
-    return weights * (function(xdata, *params) - ydata)
+def _wrap_jac(jac, xdata, weights):
+    if weights is None:
+        def jac_wrapped(params):
+            return jac(xdata, *params)
+    else:
+        def jac_wrapped(params):
+            return weights[:, np.newaxis] * jac(xdata, *params)
```

Comment:
I intentionally didn't try to cover all possible cases. People were fine without any thoughts of Jacobian for (simple) curve fitting and sparse Jacobian for curve fitting seems unusual. The main reason `jac` had to be added and handled was to eliminate potential erroneous behavior with `sigma` provided, i.e. I believe it's not good to provide something (option to pass Jacobian in kwargs) which is known to work incorrectly. 

So I don't know if it's really necessary, if you believe so I will add it. But then (in theory) we can also have `LinearOperator`. Also the description of `jac` can be extended to explain what it is and list all possible numerical schemes, i.e. basically become a copy from `least_squares`. I decided not to do it, because most likely `curve_fit` users not going to bother about it and if they want more control they better use `leastsq` or `least_squares`.